File: cli/node/src/output.ts
/**
 * Output formatting for mem0 CLI — text, JSON, table, quiet modes.
 */

import boxen from "boxen";
import Table from "cli-table3";
import { colors, sym } from "./branding.js";

const { brand, accent, success, error: errorColor, dim } = colors;

function formatDate(dtStr?: string): string | undefined {
	if (!dtStr) return undefined;
	try {
		const dt = new Date(dtStr.replace("Z", "+00:00"));
		return dt.toISOString().slice(0, 10);
	} catch {
		return dtStr?.slice(0, 10);
	}
}

export function formatMemoriesText(
	memories: Record<string, unknown>[],
	title = "memories",
): void {
	const count = memories.length;
	console.log(`\n${brand(`Found ${count} ${title}:`)}\n`);

	for (let i = 0; i < memories.length; i++) {
		const mem = memories[i];
		const memoryText = (mem.memory ?? mem.text ?? "") as string;
		const memId = ((mem.id as string) ?? "").slice(0, 8);
		const score = mem.score as number | undefined;
		const created = formatDate(mem.created_at as string | undefined);
		let category: string | undefined;
		const cats = mem.categories;
		if (Array.isArray(cats)) {
			category = cats[0] as string | undefined;
		}

		console.log(`  ${i + 1}. ${memoryText}`);

		const details: string[] = [];
		if (score !== undefined) details.push(`Score: ${score.toFixed(2)}`);
		if (memId) details.push(`ID: ${memId}`);
		if (created) details.push(`Created: ${created}`);
		if (category) details.push(`Category: ${category}`);

		if (details.length > 0) {
			console.log(`     ${dim(details.join(" · "))}`);
		}
		console.log();
	}
}

export function formatMemoriesTable(
	memories: Record<string, unknown>[],
	opts: { showScore?: boolean } = {},
): void {
	const head = opts.showScore
		? [
				accent("ID"),
				accent("Score"),
				accent("Memory"),
				accent("Category"),
				accent("Created"),
			]
		: [accent("ID"), accent("Memory"), accent("Category"), accent("Created")];
	const colWidths = opts.showScore ? [38, 8, 40, 16, 14] : [38, 40, 16, 14];
	const table = new Table({
		head,
		colWidths,
		wordWrap: true,
		style: { head: [], border: [] },
	});

	for (const mem of memories) {
		const memId = (mem.id as string) ?? "";
		let memoryText = (mem.memory ?? mem.text ?? "") as string;
		if (memoryText.length > 60) {
			memoryText = `${memoryText.slice(0, 57)}...`;
		}
		const categories = mem.categories;
		const cat =
			Array.isArray(categories) && categories.length > 0
				? categories.length > 1
					? `${categories[0]} (+${categories.length - 1})`
					: (categories[0] as string)
				: "—";
		const created = formatDate(mem.created_at as string | undefined) ?? "—";
		if (opts.showScore) {
			const score = mem.score as number | undefined;
			const scoreStr = score !== undefined ? score.toFixed(2) : "—";
			table.push([dim(memId), scoreStr, memoryText, cat, created]);
		} else {
			table.push([dim(memId), memoryText, cat, created]);
		}
	}

	console.log();
	console.log(table.toString());
	console.log();
}

export function formatJson(data: unknown): void {
	console.log(JSON.stringify(data, null, 2));
}

export function formatSingleMemory(
	mem: Record<string, unknown>,
	output = "text",
): void {
	if (output === "json") {
		formatJson(mem);
		return;
	}

	const memoryText = (mem.memory ?? mem.text ?? "") as string;
	const memId = (mem.id ?? "") as string;

	const lines: string[] = [];
	lines.push(`  ${memoryText}`);
	lines.push("");

	if (memId) lines.push(`  ${dim("ID:")}         ${memId}`);
	const created = formatDate(mem.created_at as string | undefined);
	if (created) lines.push(`  ${dim("Created:")}    ${created}`);
	const updated = formatDate(mem.updated_at as string | undefined);
	if (updated) lines.push(`  ${dim("Updated:")}    ${updated}`);
	const meta = mem.metadata;
	if (meta) lines.push(`  ${dim("Metadata:")}   ${JSON.stringify(meta)}`);
	const categories = mem.categories;
	if (categories) {
		const catStr = Array.isArray(categories)
			? categories.join(", ")
			: String(categories);
		lines.push(`  ${dim("Categories:")} ${catStr}`);
	}

	const content = lines.join("\n");
	console.log();
	console.log(
		boxen(content, {
			title: brand("Memory"),
			titleAlignment: "left",
			borderColor: "magenta",
			padding: 1,
		}),
	);
	console.log();
}

export function formatAddResult(
	result: Record<string, unknown> | Record<string, unknown>[],
	output = "text",
): void {
	if (output === "json") {
		formatJson(result);
		return;
	}
	if (output === "quiet") return;

	const results: Record<string, unknown>[] = Array.isArray(result)
		? result
		: ((result.results as Record<string, unknown>[]) ?? [result]);

	if (!results.length) {
		console.log(`  ${dim("No memories extracted.")}`);
		return;
	}

	console.log();
	const seenPendingEvents = new Set<string>();
	for (const r of results) {
		// Detect async PENDING response
		if (r.status === "PENDING") {
			const eventId = (r.event_id as string) ?? "";
			// Deduplicate PENDING entries with the same event_id
			if (eventId && seenPendingEvents.has(eventId)) continue;
			if (eventId) seenPendingEvents.add(eventId);
			const icon = accent(sym("⧗", "..."));
			const parts = [
				`  ${icon} ${dim("Queued".padEnd(10))}`,
				"Processing in background",
			];
			console.log(parts.join("  "));
			if (eventId) {
				console.log(`  ${dim(`  event_id: ${eventId}`)}`);
				console.log(
					`  ${dim(`  → Check status: mem0 event status ${eventId}`)}`,
				);
			}
			continue;
		}

		const event = (r.event ?? "ADD") as string;
		const memory = (r.memory ?? r.text ?? r.content ?? r.data ?? "") as string;
		const memId = ((r.id as string) ?? (r.memory_id as string) ?? "").slice(
			0,
			8,
		);

		let icon: string;
		let label: string;
		if (event === "ADD") {
			icon = success("+");
			label = "Added";
		} else if (event === "UPDATE") {
			icon = accent("~");
			label = "Updated";
		} else if (event === "DELETE") {
			icon = errorColor("-");
			label = "Deleted";
		} else if (event === "NOOP") {
			icon = dim("·");
			label = "No change";
		} else {
			icon = dim("?");
			label = event;
		}

		const parts = [`  ${icon} ${dim(label.padEnd(10))}`];
		if (memory) parts.push(memory);
		if (memId) parts.push(dim(`(${memId})`));
		console.log(parts.join("  "));
	}
	console.log();
}

export function formatJsonEnvelope(opts: {
	command: string;
	data: unknown;
	durationMs?: number;
	scope?: Record<string, string | undefined>;
	count?: number;
	status?: string;
	error?: string;
}): void {
	const envelope: Record<string, unknown> = {
		status: opts.status ?? "success",
		command: opts.command,
	};
	if (opts.durationMs !== undefined) envelope.duration_ms = opts.durationMs;
	if (opts.scope !== undefined) envelope.scope = opts.scope;
	if (opts.count !== undefined) envelope.count = opts.count;
	if (opts.error) envelope.error = opts.error;
	envelope.data = opts.data;
	console.log(JSON.stringify(envelope, null, 2));
}

function pick(
	obj: Record<string, unknown>,
	keys: string[],
): Record<string, unknown> {
	const result: Record<string, unknown> = {};
	for (const key of keys) {
		if (key in obj) result[key] = obj[key];
	}
	return result;
}

export function sanitizeAgentData(command: string, data: unknown): unknown {
	if (data === null || data === undefined) return data;

	switch (command) {
		case "add": {
			const items = Array.isArray(data) ? data : [data];
			return items.map((item) => {
				const r = item as Record<string, unknown>;
				if (r.status === "PENDING") return pick(r, ["status", "event_id"]);
				return pick(r, ["id", "memory", "event"]);
			});
		}
		case "search":
			return (data as Record<string, unknown>[]).map((r) =>
				pick(r, ["id", "memory", "score", "created_at", "categories"]),
			);
		case "list":
			return (data as Record<string, unknown>[]).map((r) =>
				pick(r, ["id", "memory", "created_at", "categories"]),
			);
		case "get": {
			const r = data as Record<string, unknown>;
			return pick(r, [
				"id",
				"memory",
				"created_at",
				"updated_at",
				"categories",
				"metadata",
			]);
		}
		case "update": {
			const r = data as Record<string, unknown>;
			return pick(r, ["id", "memory"]);
		}
		case "delete":
		case "delete-all":
		case "entity delete":
			return data;
		case "entity list":
			return (data as Record<string, unknown>[]).map((r) => ({
				name: (r.name ?? r.id) as string,
				...pick(r, ["type", "count"]),
			}));
		case "event list":
			return (data as Record<string, unknown>[]).map((r) =>
				pick(r, ["id", "event_type", "status", "latency", "created_at"]),
			);
		case "event status": {
			const ev = data as Record<string, unknown>;
			const rawResults =
				(ev.results as Record<string, unknown>[] | undefined) ?? [];
			const sanitizedResults = rawResults.map((r) => {
				const nested = r.data as Record<string, unknown> | undefined;
				return {
					id: r.id,
					event: r.event,
					user_id: r.user_id,
					memory: nested?.memory ?? null,
				};
			});
			return {
				...pick(ev, [
					"id",
					"event_type",
					"status",
					"latency",
					"created_at",
					"updated_at",
				]),
				results: sanitizedResults,
			};
		}
		default:
			return data;
	}
}

export function formatAgentEnvelope(opts: {
	command: string;
	data: unknown;
	durationMs?: number;
	scope?: Record<string, string | undefined>;
	count?: number;
}): void {
	const envelope: Record<string, unknown> = {
		status: "success",
		command: opts.command,
	};
	if (opts.durationMs !== undefined) envelope.duration_ms = opts.durationMs;
	if (opts.scope) {
		const filtered = Object.fromEntries(
			Object.entries(opts.scope).filter(([, v]) => v),
		);
		if (Object.keys(filtered).length > 0) envelope.scope = filtered;
	}
	if (opts.count !== undefined) envelope.count = opts.count;
	envelope.data = sanitizeAgentData(opts.command, opts.data);
	console.log(JSON.stringify(envelope, null, 2));
}

export function printResultSummary(opts: {
	count: number;
	durationSecs?: number;
	page?: number;
	scopeIds?: Record<string, string | undefined>;
}): void {
	const parts = [`${opts.count} result${opts.count !== 1 ? "s" : ""}`];
	if (opts.page !== undefined) parts.push(`page ${opts.page}`);
	if (opts.scopeIds) {
		const scopeParts = Object.entries(opts.scopeIds)
			.filter(([, v]) => v)
			.map(([k, v]) => `${k}=${v}`);
		if (scopeParts.length > 0) parts.push(scopeParts.join(", "));
	}
	if (opts.durationSecs !== undefined)
		parts.push(`${opts.durationSecs.toFixed(2)}s`);

	console.log(`  ${dim(parts.join(" · "))}`);
	console.log();
}

File: cli/python/src/mem0_cli/output.py
"""Output formatting for mem0 CLI — text, JSON, table, quiet modes."""

from __future__ import annotations

import json
from datetime import datetime
from typing import Any

from rich.console import Console
from rich.panel import Panel
from rich.table import Table
from rich.text import Text

from mem0_cli.branding import ACCENT_COLOR, BRAND_COLOR, DIM_COLOR, SUCCESS_COLOR, _sym


def format_memories_text(console: Console, memories: list[dict], title: str = "memories") -> None:
    """Render memories in human-friendly text mode."""
    count = len(memories)
    console.print(f"\n[{BRAND_COLOR}]Found {count} {title}:[/]\n")

    for i, mem in enumerate(memories, 1):
        memory_text = mem.get("memory", mem.get("text", ""))
        mem_id = mem.get("id", "")[:8]
        score = mem.get("score")
        created = _format_date(mem.get("created_at"))
        category = mem.get("categories", [None])
        if isinstance(category, list):
            category = category[0] if category else None

        line = Text()
        line.append(f"  {i}. ", style="bold")
        line.append(memory_text, style="white")
        console.print(line)

        details = []
        if score is not None:
            details.append(f"Score: {score:.2f}")
        if mem_id:
            details.append(f"ID: {mem_id}")
        if created:
            details.append(f"Created: {created}")
        if category:
            details.append(f"Category: {category}")

        if details:
            detail_str = " · ".join(details)
            console.print(f"     [{DIM_COLOR}]{detail_str}[/]")
        console.print()


def format_memories_table(
    console: Console, memories: list[dict], *, show_score: bool = False
) -> None:
    """Render memories in a rich table."""
    table = Table(
        border_style=BRAND_COLOR,
        header_style=f"bold {ACCENT_COLOR}",
        row_styles=["", "dim"],
        padding=(0, 1),
    )
    table.add_column("ID", style="dim", max_width=38, no_wrap=True)
    if show_score:
        table.add_column("Score", max_width=7, justify="right")
    table.add_column("Memory", max_width=50, no_wrap=False)
    table.add_column("Category", max_width=14)
    table.add_column("Created", max_width=12)

    for mem in memories:
        mem_id = mem.get("id", "")
        memory_text = mem.get("memory", mem.get("text", ""))
        if len(memory_text) > 60:
            memory_text = memory_text[:57] + "..."
        categories = mem.get("categories", [])
        if isinstance(categories, list) and categories:
            cat = (
                categories[0]
                if len(categories) == 1
                else f"{categories[0]} (+{len(categories) - 1})"
            )
        else:
            cat = "—"
        created = _format_date(mem.get("created_at")) or "—"
        if show_score:
            score = mem.get("score")
            score_str = f"{score:.2f}" if score is not None else "—"
            table.add_row(mem_id, score_str, memory_text, cat, created)
        else:
            table.add_row(mem_id, memory_text, cat, created)

    console.print()
    console.print(table)
    console.print()


def format_json(console: Console, data: Any) -> None:
    """Output data as pretty-printed JSON."""
    console.print_json(json.dumps(data, default=str))


def format_single_memory(console: Console, mem: dict, output: str = "text") -> None:
    """Format a single memory for display."""
    if output == "json":
        format_json(console, mem)
        return

    memory_text = mem.get("memory", mem.get("text", ""))
    mem_id = mem.get("id", "")

    lines = []
    lines.append(f"  [white bold]{memory_text}[/]")
    lines.append("")

    if mem_id:
        lines.append(f"  [{DIM_COLOR}]ID:[/]         {mem_id}")
    created = _format_date(mem.get("created_at"))
    if created:
        lines.append(f"  [{DIM_COLOR}]Created:[/]    {created}")
    updated = _format_date(mem.get("updated_at"))
    if updated:
        lines.append(f"  [{DIM_COLOR}]Updated:[/]    {updated}")
    meta = mem.get("metadata")
    if meta:
        lines.append(f"  [{DIM_COLOR}]Metadata:[/]   {json.dumps(meta)}")
    categories = mem.get("categories")
    if categories:
        cat_str = ", ".join(categories) if isinstance(categories, list) else categories
        lines.append(f"  [{DIM_COLOR}]Categories:[/] {cat_str}")

    content = "\n".join(lines)
    panel = Panel(
        content,
        title=f"[{BRAND_COLOR}]Memory[/]",
        title_align="left",
        border_style=BRAND_COLOR,
        padding=(1, 1),
    )
    console.print()
    console.print(panel)
    console.print()


def format_add_result(console: Console, result: dict | list, output: str = "text") -> None:
    """Format the result of an add operation."""
    if output == "json":
        format_json(console, result)
        return
    if output == "quiet":
        return

    # result from API is typically {"results": [...]}
    results = result if isinstance(result, list) else result.get("results", [result])
    if not results:
        console.print(f"  [{DIM_COLOR}]No memories extracted.[/]")
        return

    console.print()
    seen_pending_events: set[str] = set()
    for r in results:
        # Detect async PENDING response from Platform API
        if r.get("status") == "PENDING":
            event_id = r.get("event_id", "")
            # Deduplicate PENDING entries with the same event_id
            if event_id and event_id in seen_pending_events:
                continue
            if event_id:
                seen_pending_events.add(event_id)
            icon = f"[{ACCENT_COLOR}]{_sym('⧗', '...')}[/]"
            parts = [f"  {icon} [{DIM_COLOR}]{'Queued':<10}[/]"]
            parts.append("[white]Processing in background[/]")
            console.print("  ".join(parts))
            if event_id:
                console.print(f"  [{DIM_COLOR}]  event_id: {event_id}[/]")
                console.print(f"  [{DIM_COLOR}]  → Check status: mem0 event status {event_id}[/]")
            continue

        event = r.get("event", "ADD")
        memory = r.get("memory") or r.get("text") or r.get("content") or r.get("data") or ""
        mem_id = (r.get("id") or r.get("memory_id") or "")[:8]

        if event == "ADD":
            icon = f"[{SUCCESS_COLOR}]+[/]"
            label = "Added"
        elif event == "UPDATE":
            icon = f"[{ACCENT_COLOR}]~[/]"
            label = "Updated"
        elif event == "DELETE":
            icon = "[red]-[/]"
            label = "Deleted"
        elif event == "NOOP":
            icon = f"[{DIM_COLOR}]·[/]"
            label = "No change"
        else:
            icon = f"[{DIM_COLOR}]?[/]"
            label = event

        # Build the display line
        parts = [f"  {icon} [{DIM_COLOR}]{label:<10}[/]"]
        if memory:
            parts.append(f"[white]{memory}[/]")
        if mem_id:
            parts.append(f"[{DIM_COLOR}]({mem_id})[/]")
        console.print("  ".join(parts))
    console.print()


def format_json_envelope(
    console: Console,
    *,
    command: str,
    data: Any,
    duration_ms: int | None = None,
    scope: dict | None = None,
    count: int | None = None,
    status: str = "success",
    error: str | None = None,
) -> None:
    """Output structured JSON envelope for AI agent consumption."""
    envelope: dict[str, Any] = {
        "status": status,
        "command": command,
    }
    if duration_ms is not None:
        envelope["duration_ms"] = duration_ms
    if scope is not None:
        envelope["scope"] = scope
    if count is not None:
        envelope["count"] = count
    if error:
        envelope["error"] = error
    envelope["data"] = data

    # If the platform flagged this as an unclaimed Agent Mode account, surface
    # the notice inside the JSON envelope so an agent consuming the output
    # sees it without needing to inspect HTTP headers.
    from mem0_cli.state import take_notice

    notice = take_notice()
    if notice:
        envelope["mem0_notice"] = notice

    console.print_json(json.dumps(envelope, default=str))


def sanitize_agent_data(command: str, data: Any) -> Any:
    """Project API response data to minimal relevant fields for agent consumption."""

    def pick(obj: dict, keys: list) -> dict:
        return {k: obj[k] for k in keys if k in obj}

    if data is None:
        return data

    if command == "add":
        items = data if isinstance(data, list) else [data]
        result = []
        for item in items:
            if item.get("status") == "PENDING":
                result.append(pick(item, ["status", "event_id"]))
            else:
                result.append(pick(item, ["id", "memory", "event"]))
        return result

    if command == "search":
        return [pick(r, ["id", "memory", "score", "created_at", "categories"]) for r in data]

    if command == "list":
        return [pick(r, ["id", "memory", "created_at", "categories"]) for r in data]

    if command == "get":
        return pick(data, ["id", "memory", "created_at", "updated_at", "categories", "metadata"])

    if command == "update":
        return pick(data, ["id", "memory"])

    if command in ("delete", "delete-all", "entity delete"):
        return data

    if command == "entity list":
        result = []
        for r in data:
            item = pick(r, ["type", "count"])
            item["name"] = r.get("name") or r.get("id", "")
            result.append(item)
        return result

    if command == "event list":
        return [pick(r, ["id", "event_type", "status", "latency", "created_at"]) for r in data]

    if command == "event status":
        ev = data
        raw_results = ev.get("results") or []
        sanitized_results = []
        for r in raw_results:
            nested = r.get("data") or {}
            memory = nested.get("memory") if isinstance(nested, dict) else None
            sanitized_results.append(
                {
                    "id": r.get("id"),
                    "event": r.get("event"),
                    "user_id": r.get("user_id"),
                    "memory": memory,
                }
            )
        result = pick(ev, ["id", "event_type", "status", "latency", "created_at", "updated_at"])
        result["results"] = sanitized_results
        return result

    # Pass-through: status, import, config show/get/set
    return data


def format_agent_envelope(
    console: Console,
    *,
    command: str,
    data: Any,
    duration_ms: int | None = None,
    scope: dict | None = None,
    count: int | None = None,
) -> None:
    """Output structured JSON envelope for agent/programmatic use (--json/--agent mode)."""
    envelope: dict[str, Any] = {
        "status": "success",
        "command": command,
    }
    if duration_ms is not None:
        envelope["duration_ms"] = duration_ms
    if scope:
        filtered = {k: v for k, v in scope.items() if v}
        if filtered:
            envelope["scope"] = filtered
    if count is not None:
        envelope["count"] = count
    envelope["data"] = sanitize_agent_data(command, data)

    # Surface the unclaimed-Agent-Mode notice (if any) in the envelope so an
    # agent reading the JSON output sees it without inspecting HTTP headers.
    from mem0_cli.state import take_notice

    notice = take_notice()
    if notice:
        envelope["mem0_notice"] = notice

    console.print_json(json.dumps(envelope, default=str))


def print_result_summary(
    console: Console,
    count: int,
    *,
    duration_secs: float | None = None,
    page: int | None = None,
    **scope_ids: str | None,
) -> None:
    """Print a summary footer after result lists."""
    parts = [f"{count} result{'s' if count != 1 else ''}"]
    if page is not None:
        parts.append(f"page {page}")
    scope_parts = [f"{k}={v}" for k, v in scope_ids.items() if v]
    if scope_parts:
        parts.append(", ".join(scope_parts))
    if duration_secs is not None:
        parts.append(f"{duration_secs:.2f}s")

    summary = " · ".join(parts)
    console.print(f"  [{DIM_COLOR}]{summary}[/]")
    console.print()


def _format_date(dt_str: str | None) -> str | None:
    if not dt_str:
        return None
    try:
        dt = datetime.fromisoformat(dt_str.replace("Z", "+00:00"))
        return dt.strftime("%Y-%m-%d")
    except (ValueError, AttributeError):
        return str(dt_str)[:10] if dt_str else None
